File: code/java/ch10/practice/Warmup01Factorial.java
package ch10.practice;

import java.util.*;

/**
 * Warmup 01: Factorial
 * ==============================
 * Chapter 10: The Magic of Recursion
 *
 * PROBLEM: Given a non-negative integer n, return n! (n factorial).
 *          n! = n * (n-1) * ... * 1, with 0! = 1.
 *
 * ALGORITHM: Use recursion with base case n == 0.
 *
 * EXAMPLES:
 *   solve(0)  = 1
 *   solve(1)  = 1
 *   solve(5)  = 120
 *   solve(10) = 3628800
 *
 * CONSTRAINTS: 0 <= n <= 20 (result fits in long)
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Warmup01Factorial {
    public static long solve(int n) {
        // TODO: Replace this with your solution
        return 0;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(n));
        sc.close();
    }
}

File: code/python/ch10/practice/warmup_01_factorial.py
"""
Warmup 1: Factorial
==============================
Chapter 10: The Magic of Recursion — Functions That Call Themselves

PROBLEM
-------
Compute n! (n factorial) using recursion.

n! is defined as:
  0! = 1
  n! = n * (n-1)! for n > 0

INPUT FORMAT
------------
A single integer n.

OUTPUT FORMAT
-------------
A single integer — the value of n!.

CONSTRAINTS
-----------
- 0 <= n <= 20

EXAMPLES
--------
Input:
  5
Output: 120

Input:
  0
Output: 1

Input:
  1
Output: 1

HINT
----
The base case is n == 0 (returns 1). For the recursive case,
multiply n by the factorial of n - 1.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(n: int) -> int:
    """Compute n! recursively."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    n = int(input())
    print(solve(n))
